github/vscode-codeql
# Import testproj databases into workspace storage

Commit-by-commit review is recommended.



If a folder that ends with `.testproj` is encountered, assume it is a
database created by a codeql test. When the user wants to import this
database, copy it into workspace storage. The database can be
re-imported, which first removes the old version before importing it
again.

Ask user if they want to re-import outdated testproj dbs

Before running a query now, do the following:

1. Check if the selected database is imported from a testproj
2. If so, check the last modified time of the `codeql-datase.yml` file
   of the imported database with that of its origin.
3. If the origin database has a file that is newer, assume that the
   database has been recreated since the last time it was imported.
4. If newer, then ask the user if they want to re-import before running
   the query.

Also, this change appends the `(test)` label to all test databases in
the database list.

To verify this change, do the following:

1. In the development workspace directory, create a codeql test:
    ```
    echo "import javascript
    select 1" | tee  test.ql
    echo "dependencies:
        codeql/javascript-all: \"*\"
    extractor: javascript
    tests: test"
    echo hucairz | tee qlpack.yml
    ```
2. Start the vscode dev workbench
3. In the vscode editor, go to the test tab (looks like a beaker)
4. Run the test. It should fail
5. Right click on the `test.testproj` directory and select "(Re-)Import test database"
6. Notice that the database is imported
7. Run the query if you like and you should see `1` as the result.
9. Change test to `select 2`
10. Re-run the test
11. Re-run the query
12. This time, you should get a popup asking if you want to re-import
13. Select `yes`
14. See the results are now `2`.

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.
